File: Golang/Easy/1672_richest_customer_wealth.go
package main

import "fmt"

func sum(list []int) int {
	ans := 0
	for _, val := range list {
		ans += val
	}
	return ans
}

// maximumWealth returns the largest row-sum in accounts — i.e., the wealth of
// the richest customer.
//
// Sum each row and keep the running max. No materialized totals needed.
//
// Time complexity:  O(m * n)
// Space complexity: O(1)
func maximumWealth(accounts [][]int) int {
	ans := 0

	for _, acc := range accounts {
		ans = max(sum(acc), ans)
	}
	return ans
}

func main() {
	testCases := []struct {
		accounts [][]int
		expected int
	}{
		{[][]int{{1, 2, 3}, {3, 2, 1}}, 6},
		{[][]int{{1, 5}, {7, 3}, {3, 5}}, 10},
		{[][]int{{2, 8, 7}, {7, 1, 3}, {1, 9, 5}}, 17},
		{[][]int{{100}}, 100},
	}

	for i, tc := range testCases {
		result := maximumWealth(tc.accounts)
		status := "PASS"
		if result != tc.expected {
			status = "FAIL"
		}
		fmt.Printf("Example %d | Input: %v | Expected: %d | Result: %d | [%s]\n",
			i+1, tc.accounts, tc.expected, result, status)
	}
}

File: Python/Easy/1672_richest_customer_wealth.py
"""
1672. Richest Customer Wealth

Given an m x n grid accounts where accounts[i][j] is customer i's balance at
bank j, return the maximum total wealth across all customers.

Problem Link: https://leetcode.com/problems/richest-customer-wealth/

Examples:
    Example 1:
        Input: accounts = [[1, 2, 3], [3, 2, 1]]
        Output: 6

    Example 2:
        Input: accounts = [[1, 5], [7, 3], [3, 5]]
        Output: 10

    Example 3:
        Input: accounts = [[2, 8, 7], [7, 1, 3], [1, 9, 5]]
        Output: 17

Constraints:
    m == accounts.length
    n == accounts[i].length
    1 <= m, n <= 50
    1 <= accounts[i][j] <= 100
"""


class Solution:
    def maximumWealth(self, accounts: list[list[int]]) -> int:
        """
        Sum each customer's row and track the running max. Single pass over
        the grid — no need to materialize the per-customer totals.

        Time complexity:  O(m * n)
        Space complexity: O(1)
        """
        ans: int = 0
        for acc in accounts:
            ans = max(sum(acc), ans)
        return ans


if __name__ == "__main__":
    solution = Solution()

    test_cases = [
        ([[1, 2, 3], [3, 2, 1]], 6),
        ([[1, 5], [7, 3], [3, 5]], 10),
        ([[2, 8, 7], [7, 1, 3], [1, 9, 5]], 17),
        ([[100]], 100),
    ]

    for i, (accounts, expected) in enumerate(test_cases, 1):
        result = solution.maximumWealth(accounts)
        assert result == expected, f"Test {i} failed: expected {expected}, got {result}"

    print("All test cases passed!")
